tests/src/test/kotlin/dev/restate/e2e/CancelInvocationTest.kt

```diff
@@ -0,0 +1,129 @@
+import org.junit.jupiter.params.provider.MethodSource
+
+class JavaCancelInvocationTest : BaseCancelInvocationTest() {
+  companion object {
+    @RegisterExtension
+    val deployerExt: RestateDeployerExtension =
+        RestateDeployerExtension(
+            RestateDeployer.Builder()
+                .withServiceEndpoint(
+                    Containers.javaServicesContainer(
+                        "java-cancel-invocation",
+                        CancelTestServiceGrpc.SERVICE_NAME,
+                        BlockingServiceGrpc.SERVICE_NAME))
+                .withServiceEndpoint(
+                    Containers.nodeServicesContainer(
+                        "awakeable-holder", AwakeableHolderServiceGrpc.SERVICE_NAME))
+                .build())
+  }
+}
+
+class NodeCancelInvocationTest : BaseCancelInvocationTest() {
+  companion object {
+    @RegisterExtension
+    val deployerExt: RestateDeployerExtension =
+        RestateDeployerExtension(
+            RestateDeployer.Builder()
+                .withServiceEndpoint(
+                    Containers.nodeServicesContainer(
+                        "node-cancel-invocation",
+                        CancelTestServiceGrpc.SERVICE_NAME,
+                        BlockingServiceGrpc.SERVICE_NAME,
+                        AwakeableHolderServiceGrpc.SERVICE_NAME))
+                .build())
+  }
+}
+
+abstract class BaseCancelInvocationTest {
+  @ParameterizedTest(name = "cancel blocked invocation on {0}")
+  @MethodSource("blockingOperations")
+  fun cancelInvocation(
+      blockingOperation: BlockingOperation,
+      @InjectBlockingStub ingressClient: IngressBlockingStub,
+      @InjectChannel channel: Channel,
+      @InjectBlockingStub blockingServiceClient: BlockingServiceGrpc.BlockingServiceBlockingStub,
+      @InjectBlockingStub
+      awakeableHolderClient: AwakeableHolderServiceGrpc.AwakeableHolderServiceBlockingStub,
+      @InjectMetaURL metaURL: URL,
+  ) {
+    val request = blockingRequest { operation = blockingOperation }
+
+    val id =
+        ingressClient
+            .invoke(
+                invokeRequest {
+                  service = CancelTestServiceGrpc.SERVICE_NAME
+                  method = CancelTestServiceGrpc.getStartTestMethod().bareMethodName!!
+                  pb = request.toByteString()
+                })
+            .id
+
+    await untilCallTo
+        {
+          awakeableHolderClient.hasAwakeable(hasAwakeableRequest { name = "cancel" })
+        } matches
+        { result ->
+          result!!.hasAwakeable
+        }
+
+    awakeableHolderClient.unlock(unlockRequest { name = "cancel" })
+
+    val client = InvocationApi(ApiClient().setHost(metaURL.host).setPort(metaURL.port))
+    val cancelTestClient = CancelTestServiceGrpcKt.CancelTestServiceCoroutineStub(channel)
+
+    // The termination signal might arrive before the blocking call to the cancel singleton was
+    // made,
+    // so we need to retry.
```

Comment:
You can add a sync point with AwakeableHolder to CancelTestService as well.

Reply:
This wouldn't work for example when sleeping in the `BlockingService`. Will keep it as is.